src/htr2hpc/api_client.py

```diff
@@ -0,0 +1,329 @@
+import datetime
+import logging
+from collections import namedtuple
+from dataclasses import dataclass
+from time import sleep
+import pathlib
+
+import humanize
+import requests
+from django.utils.text import slugify  # is django a reasonable dependency?
+
+from htr2hpc import __version__ as _version
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ResultsList:
+    """API list response."""
+
+    # all API list methods have the same structure,
+    # so use a dataclass but specify the result type
+
+    result_type: str
+    count: int
+    next: str
+    previous: str
+    results: list
+
+    def __post_init__(self):
+        # convert result entries to namedtuple class based on
+        # specified result type name
+        self.results = [to_namedtuple(self.result_type, d) for d in self.results]
+
+
+@dataclass
+class Task:
+    """API response for a task result."""
+
+    pk: int
+    document: int
+    document_part: int
+    workflow_state: int
+    label: str
+    messages: str
+    queued_at: datetime.datetime
+    started_at: datetime.datetime
+    done_at: datetime.datetime
+    method: str
+    user: int
+
+    # NOTE: workflow state is a numeric id defined in escriptorium
+    # current values:
+    # WORKFLOW_STATE_QUEUED = 0
+    # WORKFLOW_STATE_STARTED = 1
+    # WORKFLOW_STATE_ERROR = 2
+    # WORKFLOW_STATE_DONE = 3
+    # WORKFLOW_STATE_CANCELED = 4
+    # add a property?
+
+    def __post_init__(self):
+        # convert dates from string to datetime
+        for date_field in ["queued_at", "started_at", "done_at"]:
+            value = getattr(self, date_field)
+            if value:
+                setattr(self, date_field, datetime.datetime.fromisoformat(value))
+
+    def duration(self) -> datetime.timedelta | None:
+        "how long the task took to complete"
+        if self.done_at is not None:
+            return self.done_at - self.started_at
+
+
+# keep a registry of result classes for API result objects,
+# so they can be reused once they are defined
```

Comment:
How varied are these result classes? Do these change enough that it's not possible to define them upfront?